src/Angor/Avalonia/Angor.Contexts.Wallet/Infrastructure/Impl/History/TransactionHistory.cs

```diff
@@ -29,78 +26,67 @@ public Task<Result<IEnumerable<string>>> GetWalletAddresses(WalletWords walletWo
+        var result = await GetWalletAddresses(walletWords);
+        if (result.IsFailure)
+        {
+            return Result.Failure<IEnumerable<BroadcastedTransaction>>(result.Error);
+        }
+        return await result
+            .Bind(addresses => addresses
+                .Select(s => LookupAddressTransactions(s))
+                .Combine())
+            .Map(transactions => transactions
+                .SelectMany(t => t))
+            .Bind(transactions =>
+            {
+                var addresses = result.Value.Select(x => new Address(x)).ToList();
+                return transactions
+                    .Select(tx => MapToBroadcastedTransaction(tx, addresses))
+                    .Combine();
+            });
+
     }
 
-    private Task<Result<BroadcastedTransaction>> CreateBroadcastedTransaction(QueryTransaction tx, WalletWords walletWords)
+    private static Task<Result<BroadcastedTransaction>> MapToBroadcastedTransaction(QueryTransaction tx,
+        List<Address> walletAddresses)
     {
-        var (inputs, outputs) = MapInputsAndOutputs(tx);
+        var inputs = MapToUiModelInputs(tx.Inputs);
+        var outputs = MapToUiModelOutputs(tx.Outputs);
 
-        return GetWalletAddresses(walletWords)
-            .Map(enumerable => enumerable.Select(a => new Address(a)))
-            .Map(walletAddresses =>
-        {
-            return new BroadcastedTransaction(
-                Id: tx.TransactionId,
-                WalletInputs: inputs.Where(i => walletAddresses.Contains(i.Address)),
-                WalletOutputs: outputs.Where(o => walletAddresses.Contains(o.Address)),
-                AllInputs: inputs,
-                AllOutputs: outputs,
-                Fee: tx.Fee,
-                IsConfirmed: tx.Confirmations > 0,
-                BlockHeight: tx.BlockIndex,
-                BlockTime: tx.Timestamp > 0 ? DateTimeOffset.FromUnixTimeSeconds(tx.Timestamp) : null,
-                RawJson: JsonSerializer.Serialize(tx)
-            );
-        });
+        return Task.FromResult(Result.Try(() => new BroadcastedTransaction(
+            Id: tx.TransactionId,
+            WalletInputs: inputs
+                .Where(i => walletAddresses.Contains(i.Address)),
+            WalletOutputs: outputs.Where(o => walletAddresses.Contains(o.Address)),
+            AllInputs: inputs,
+            AllOutputs: outputs,
+            Fee: tx.Fee,
+            IsConfirmed: tx.Confirmations > 0,
+            BlockHeight: tx.BlockIndex,
+            BlockTime: tx.Timestamp > 0 ? DateTimeOffset.FromUnixTimeSeconds(tx.Timestamp) : null,
+            RawJson: JsonSerializer.Serialize(tx)
+        )));
     }
 
-    private static (List<TransactionInput> inputs, List<TransactionOutput> outputs) MapInputsAndOutputs(QueryTransaction tx)
+    private static List<TransactionInput> MapToUiModelInputs(IEnumerable<QueryTransactionInput> inputs)
```

Comment:
Hey, noticed the MapToUIModelInputs method name in your code. In DDD, infrastructure services shouldn't reference UI directly. Could you rename it to something like **MapToTransactionInputs** to maintain proper layer separation? Thanks!